examples/seismic/source.py

```diff
@@ -1,60 +1,153 @@
+
+        self.start = start
+        self.stop = stop
+        self.step = step
+        self.num = num
+
+    def __str__(self):
+        return "TimeAxis: start=%g, stop=%g, step=%g, num=%g" % \
+               (self.start, self.stop, self.step, self.num)
+
+    @cached_property
+    def time_values(self):
+        return np.linspace(self.start, self.stop, self.num)
 
 
 class PointSource(SparseTimeFunction):
     """Symbolic data object for a set of sparse point sources
 
-    :param name: Name of the symbol representing this source
+    :param name: Name of the symbol representing this source.
     :param grid: :class:`Grid` object defining the computational domain.
-    :param coordinates: Point coordinates for this source
-    :param data: (Optional) Data values to initialise point data
-    :param ntime: (Optional) Number of timesteps for which to allocate data
-    :param npoint: (Optional) Number of sparse points represented by this source
-    :param time_order: (Optional) Time discretization order (defaults to 2)
+    :param coordinates: Point coordinates for this source.
+    :param time_range: :class:`TimeAxis` TimeAxis(start, step, num) object.
+    :param data: (Optional) Data values to initialise point data.
+    :param npoint: (Optional) Number of sparse points represented by this source.
+    :param time_order: (Optional) Time discretization order (defaults to 2).
     :param dimension: :(Optional) class:`Dimension` object for
-                       representing the number of points in this source
-
-    Note, either the dimensions `ntime` and `npoint` or the fully
-    initialised `data` array need to be provided.
+                       representing the number of points in this source.
     """
 
-    def __new__(cls, name, grid, ntime=None, npoint=None, data=None,
-                coordinates=None, **kwargs):
+    def __new__(cls, name, grid, time_range, npoint=None,
+                data=None, coordinates=None, **kwargs):
         p_dim = kwargs.get('dimension', Dimension(name='p_%s' % name))
         time_order = kwargs.get('time_order', 2)
         npoint = npoint or coordinates.shape[0]
-        if data is None:
-            if ntime is None:
-                error('Either data or ntime are required to'
-                      'initialise source/receiver objects')
-        else:
-            ntime = ntime or data.shape[0]
 
         # Create the underlying SparseTimeFunction object
         obj = SparseTimeFunction.__new__(cls, name=name, grid=grid,
                                          dimensions=[grid.time_dim, p_dim],
-                                         npoint=npoint, nt=ntime, time_order=time_order,
+                                         npoint=npoint, nt=time_range.num,
+                                         time_order=time_order,
                                          coordinates=coordinates, **kwargs)
 
+        obj._time_range = copy(time_range)
```

Comment:
nitpicking: I'd prefer a ``_reconstruct`` method, because that way you have full control on what is being copied and what not. It'd also be closer to what I do throughout Devito. See for example https://github.com/opesci/devito/blob/master/devito/ir/support/space.py#L124

Reply:
fixed in ca4d5203